File: golang/github.go
// To parse and unparse this JSON data, add this code to your project and do:
//
//    aPIData, err := UnmarshalAPIData(bytes)
//    bytes, err = aPIData.Marshal()
//
//    emojis, err := UnmarshalEmojis(bytes)
//    bytes, err = emojis.Marshal()
//
//    events, err := UnmarshalEvents(bytes)
//    bytes, err = events.Marshal()
//
//    gists, err := UnmarshalGists(bytes)
//    bytes, err = gists.Marshal()
//
//    meta, err := UnmarshalMeta(bytes)
//    bytes, err = meta.Marshal()

package main

import "encoding/json"

type APIData map[string]string

func UnmarshalAPIData(data []byte) (APIData, error) {
	var r APIData
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *APIData) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

type Emojis map[string]string

func UnmarshalEmojis(data []byte) (Emojis, error) {
	var r Emojis
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *Emojis) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

type Events []Event

func UnmarshalEvents(data []byte) (Events, error) {
	var r Events
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *Events) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

type Gists []Gist

func UnmarshalGists(data []byte) (Gists, error) {
	var r Gists
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *Gists) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

func UnmarshalMeta(data []byte) (Meta, error) {
	var r Meta
	err := json.Unmarshal(data, &r)
	return r, err
}

func (r *Meta) Marshal() ([]byte, error) {
	return json.Marshal(r)
}

type Event struct {
	ID        string    `json:"id"`           
	Type      string    `json:"type"`         
	Actor     Actor     `json:"actor"`        
	Repo      EventRepo `json:"repo"`         
	Payload   Payload   `json:"payload"`      
	Public    bool      `json:"public"`       
	CreatedAt string    `json:"created_at"`   
	Org       *Actor    `json:"org,omitempty"`
}

type Actor struct {
	ID           int64   `json:"id"`                     
	Login        string  `json:"login"`                  
	DisplayLogin *string `json:"display_login,omitempty"`
	GravatarID   string  `json:"gravatar_id"`            
	URL          string  `json:"url"`                    
	AvatarURL    string  `json:"avatar_url"`             
}

type Payload struct {
	PushID       *int64       `json:"push_id,omitempty"`      
	Size         *int64       `json:"size,omitempty"`         
	DistinctSize *int64       `json:"distinct_size,omitempty"`
	Ref          *string      `json:"ref"`                    
	Head         *string      `json:"head,omitempty"`         
	Before       *string      `json:"before,omitempty"`       
	Commits      []Commit     `json:"commits"`                
	Action       *string      `json:"action,omitempty"`       
	Number       *int64       `json:"number,omitempty"`       
	PullRequest  *PullRequest `json:"pull_request,omitempty"` 
	RefType      *string      `json:"ref_type,omitempty"`     
	MasterBranch *string      `json:"master_branch,omitempty"`
	Description  *string      `json:"description"`            
	PusherType   *string      `json:"pusher_type,omitempty"`  
	Comment      *Comment     `json:"comment,omitempty"`      
}

type Comment struct {
	URL                 string       `json:"url"`                   
	PullRequestReviewID int64        `json:"pull_request_review_id"`
	ID                  int64        `json:"id"`                    
	NodeID              string       `json:"node_id"`               
	DiffHunk            string       `json:"diff_hunk"`             
	Path                string       `json:"path"`                  
	Position            int64        `json:"position"`              
	OriginalPosition    int64        `json:"original_position"`     
	CommitID            string       `json:"commit_id"`             
	OriginalCommitID    string       `json:"original_commit_id"`    
	User                Owner        `json:"user"`                  
	Body                string       `json:"body"`                  
	CreatedAt           string       `json:"created_at"`            
	UpdatedAt           string       `json:"updated_at"`            
	HTMLURL             string       `json:"html_url"`              
	PullRequestURL      string       `json:"pull_request_url"`      
	AuthorAssociation   string       `json:"author_association"`    
	Links               CommentLinks `json:"_links"`                
}

type CommentLinks struct {
	Self        HTML `json:"self"`        
	HTML        HTML `json:"html"`        
	PullRequest HTML `json:"pull_request"`
}

type HTML struct {
	Href string `json:"href"`
}

type Owner struct {
	Login             string `json:"login"`              
	ID                int64  `json:"id"`                 
	NodeID            string `json:"node_id"`            
	AvatarURL         string `json:"avatar_url"`         
	GravatarID        string `json:"gravatar_id"`        
	URL               string `json:"url"`                
	HTMLURL           string `json:"html_url"`           
	FollowersURL      string `json:"followers_url"`      
	FollowingURL      string `json:"following_url"`      
	GistsURL          string `json:"gists_url"`          
	StarredURL        string `json:"starred_url"`        
	SubscriptionsURL  string `json:"subscriptions_url"`  
	OrganizationsURL  string `json:"organizations_url"`  
	ReposURL          string `json:"repos_url"`          
	EventsURL         string `json:"events_url"`         
	ReceivedEventsURL string `json:"received_events_url"`
	Type              Type   `json:"type"`               
	SiteAdmin         bool   `json:"site_admin"`         
}

type Commit struct {
	SHA      string `json:"sha"`     
	Author   Author `json:"author"`  
	Message  string `json:"message"` 
	Distinct bool   `json:"distinct"`
	URL      string `json:"url"`     
}

type Author struct {
	Email string `json:"email"`
	Name  string `json:"name"` 
}

type PullRequest struct {
	URL                 string           `json:"url"`                            
	ID                  int64            `json:"id"`                             
	NodeID              string           `json:"node_id"`                        
	HTMLURL             string           `json:"html_url"`                       
	DiffURL             string           `json:"diff_url"`                       
	PatchURL            string           `json:"patch_url"`                      
	IssueURL            string           `json:"issue_url"`                      
	Number              int64            `json:"number"`                         
	State               string           `json:"state"`                          
	Locked              bool             `json:"locked"`                         
	Title               string           `json:"title"`                          
	User                Owner            `json:"user"`                           
	Body                *string          `json:"body"`                           
	CreatedAt           string           `json:"created_at"`                     
	UpdatedAt           string           `json:"updated_at"`                     
	ClosedAt            interface{}      `json:"closed_at"`                      
	MergedAt            interface{}      `json:"merged_at"`                      
	MergeCommitSHA      *string          `json:"merge_commit_sha"`               
	Assignee            interface{}      `json:"assignee"`                       
	Assignees           []interface{}    `json:"assignees"`                      
	RequestedReviewers  []interface{}    `json:"requested_reviewers"`            
	RequestedTeams      []interface{}    `json:"requested_teams"`                
	Labels              []interface{}    `json:"labels"`                         
	Milestone           interface{}      `json:"milestone"`                      
	CommitsURL          string           `json:"commits_url"`                    
	ReviewCommentsURL   string           `json:"review_comments_url"`            
	ReviewCommentURL    string           `json:"review_comment_url"`             
	CommentsURL         string           `json:"comments_url"`                   
	StatusesURL         string           `json:"statuses_url"`                   
	Head                Base             `json:"head"`                           
	Base                Base             `json:"base"`                           
	Links               PullRequestLinks `json:"_links"`                         
	AuthorAssociation   string           `json:"author_association"`             
	Merged              *bool            `json:"merged,omitempty"`               
	Mergeable           interface{}      `json:"mergeable"`                      
	Rebaseable          interface{}      `json:"rebaseable"`                     
	MergeableState      *string          `json:"mergeable_state,omitempty"`      
	MergedBy            interface{}      `json:"merged_by"`                      
	Comments            *int64           `json:"comments,omitempty"`             
	ReviewComments      *int64           `json:"review_comments,omitempty"`      
	MaintainerCanModify *bool            `json:"maintainer_can_modify,omitempty"`
	Commits             *int64           `json:"commits,omitempty"`              
	Additions           *int64           `json:"additions,omitempty"`            
	Deletions           *int64           `json:"deletions,omitempty"`            
	ChangedFiles        *int64           `json:"changed_files,omitempty"`        
}

type Base struct {
	Label string   `json:"label"`
	Ref   string   `json:"ref"`  
	SHA   string   `json:"sha"`  
	User  Owner    `json:"user"` 
	Repo  BaseRepo `json:"repo"` 
}

type BaseRepo struct {
	ID               int64       `json:"id"`               
	NodeID           string      `json:"node_id"`          
	Name             string      `json:"name"`             
	FullName         string      `json:"full_name"`        
	Owner            Owner       `json:"owner"`            
	Private          bool        `json:"private"`          
	HTMLURL          string      `json:"html_url"`         
	Description      *string     `json:"description"`      
	Fork             bool        `json:"fork"`             
	URL              string      `json:"url"`              
	ForksURL         string      `json:"forks_url"`        
	KeysURL          string      `json:"keys_url"`         
	CollaboratorsURL string      `json:"collaborators_url"`
	TeamsURL         string      `json:"teams_url"`        
	HooksURL         string      `json:"hooks_url"`        
	IssueEventsURL   string      `json:"issue_events_url"` 
	EventsURL        string      `json:"events_url"`       
	AssigneesURL     string      `json:"assignees_url"`    
	BranchesURL      string      `json:"branches_url"`     
	TagsURL          string      `json:"tags_url"`         
	BlobsURL         string      `json:"blobs_url"`        
	GitTagsURL       string      `json:"git_tags_url"`     
	GitRefsURL       string      `json:"git_refs_url"`     
	TreesURL         string      `json:"trees_url"`        
	StatusesURL      string      `json:"statuses_url"`     
	LanguagesURL     string      `json:"languages_url"`    
	StargazersURL    string      `json:"stargazers_url"`   
	ContributorsURL  string      `json:"contributors_url"` 
	SubscribersURL   string      `json:"subscribers_url"`  
	SubscriptionURL  string      `json:"subscription_url"` 
	CommitsURL       string      `json:"commits_url"`      
	GitCommitsURL    string      `json:"git_commits_url"`  
	CommentsURL      string      `json:"comments_url"`     
	IssueCommentURL  string      `json:"issue_comment_url"`
	ContentsURL      string      `json:"contents_url"`     
	CompareURL       string      `json:"compare_url"`      
	MergesURL        string      `json:"merges_url"`       
	ArchiveURL       string      `json:"archive_url"`      
	DownloadsURL     string      `json:"downloads_url"`    
	IssuesURL        string      `json:"issues_url"`       
	PullsURL         string      `json:"pulls_url"`        
	MilestonesURL    string      `json:"milestones_url"`   
	NotificationsURL string      `json:"notifications_url"`
	LabelsURL        string      `json:"labels_url"`       
	ReleasesURL      string      `json:"releases_url"`     
	DeploymentsURL   string      `json:"deployments_url"`  
	CreatedAt        string      `json:"created_at"`       
	UpdatedAt        string      `json:"updated_at"`       
	PushedAt         string      `json:"pushed_at"`        
	GitURL           string      `json:"git_url"`          
	SSHURL           string      `json:"ssh_url"`          
	CloneURL         string      `json:"clone_url"`        
	SvnURL           string      `json:"svn_url"`          
	Homepage         *string     `json:"homepage"`         
	Size             int64       `json:"size"`             
	StargazersCount  int64       `json:"stargazers_count"` 
	WatchersCount    int64       `json:"watchers_count"`   
	Language         string      `json:"language"`         
	HasIssues        bool        `json:"has_issues"`       
	HasProjects      bool        `json:"has_projects"`     
	HasDownloads     bool        `json:"has_downloads"`    
	HasWiki          bool        `json:"has_wiki"`         
	HasPages         bool        `json:"has_pages"`        
	ForksCount       int64       `json:"forks_count"`      
	MirrorURL        interface{} `json:"mirror_url"`       
	Archived         bool        `json:"archived"`         
	OpenIssuesCount  int64       `json:"open_issues_count"`
	License          *License    `json:"license"`          
	Forks            int64       `json:"forks"`            
	OpenIssues       int64       `json:"open_issues"`      
	Watchers         int64       `json:"watchers"`         
	DefaultBranch    string      `json:"default_branch"`   
}

type License struct {
	Key    string  `json:"key"`    
	Name   string  `json:"name"`   
	SpdxID *string `json:"spdx_id"`
	URL    *string `json:"url"`    
	NodeID string  `json:"node_id"`
}

type PullRequestLinks struct {
	Self           HTML `json:"self"`           
	HTML           HTML `json:"html"`           
	Issue          HTML `json:"issue"`          
	Comments       HTML `json:"comments"`       
	ReviewComments HTML `json:"review_comments"`
	ReviewComment  HTML `json:"review_comment"` 
	Commits        HTML `json:"commits"`        
	Statuses       HTML `json:"statuses"`       
}

type EventRepo struct {
	ID   int64  `json:"id"`  
	Name string `json:"name"`
	URL  string `json:"url"` 
}

type Gist struct {
	URL         string          `json:"url"`         
	ForksURL    string          `json:"forks_url"`   
	CommitsURL  string          `json:"commits_url"` 
	ID          string          `json:"id"`          
	NodeID      string          `json:"node_id"`     
	GitPullURL  string          `json:"git_pull_url"`
	GitPushURL  string          `json:"git_push_url"`
	HTMLURL     string          `json:"html_url"`    
	Files       map[string]File `json:"files"`       
	Public      bool            `json:"public"`      
	CreatedAt   string          `json:"created_at"`  
	UpdatedAt   string          `json:"updated_at"`  
	Description *string         `json:"description"` 
	Comments    int64           `json:"comments"`    
	User        interface{}     `json:"user"`        
	CommentsURL string          `json:"comments_url"`
	Owner       Owner           `json:"owner"`       
	Truncated   bool            `json:"truncated"`   
}

type File struct {
	Filename string  `json:"filename"`
	Type     string  `json:"type"`    
	Language *string `json:"language"`
	RawURL   string  `json:"raw_url"` 
	Size     int64   `json:"size"`    
}

type Meta struct {
	VerifiablePasswordAuthentication bool     `json:"verifiable_password_authentication"`
	GithubServicesSHA                string   `json:"github_services_sha"`               
	Hooks                            []string `json:"hooks"`                             
	Git                              []string `json:"git"`                               
	Pages                            []string `json:"pages"`                             
	Importer                         []string `json:"importer"`                          
}

type Type string
const (
	Organization Type = "Organization"
	User Type = "User"
)

File: kotlin/github.kt
// To parse the JSON, install Klaxon and do:
//
//   val aPIData = APIData.fromJson(jsonString)
//   val emojis = Emojis.fromJson(jsonString)
//   val events = Events.fromJson(jsonString)
//   val gists = Gists.fromJson(jsonString)
//   val meta = Meta.fromJson(jsonString)

package quicktype

import com.beust.klaxon.*

private fun <T> Klaxon.convert(k: kotlin.reflect.KClass<*>, fromJson: (JsonValue) -> T, toJson: (T) -> String, isUnion: Boolean = false) =
    this.converter(object: Converter {
        @Suppress("UNCHECKED_CAST")
        override fun toJson(value: Any)        = toJson(value as T)
        override fun fromJson(jv: JsonValue)   = fromJson(jv) as Any
        override fun canConvert(cls: Class<*>) = cls == k.java || (isUnion && cls.superclass == k.java)
    })

private val klaxon = Klaxon()
    .convert(Type::class, { Type.fromValue(it.string!!) }, { "\"${it.value}\"" })

class APIData(elements: Map<String, String>) : HashMap<String, String>(elements) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = APIData (
            klaxon.parseJsonObject(java.io.StringReader(json)) as Map<String, String>
        )
    }
}
class Emojis(elements: Map<String, String>) : HashMap<String, String>(elements) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = Emojis (
            klaxon.parseJsonObject(java.io.StringReader(json)) as Map<String, String>
        )
    }
}
class Events(elements: Collection<Event>) : ArrayList<Event>(elements) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = Events(klaxon.parseArray<Event>(json)!!)
    }
}
class Gists(elements: Collection<Gist>) : ArrayList<Gist>(elements) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = Gists(klaxon.parseArray<Gist>(json)!!)
    }
}

data class Event (
    val id: String,
    val type: String,
    val actor: Actor,
    val repo: EventRepo,
    val payload: Payload,
    val public: Boolean,

    @Json(name = "created_at")
    val createdAt: String,

    val org: Actor? = null
)

data class Actor (
    val id: Long,
    val login: String,

    @Json(name = "display_login")
    val displayLogin: String? = null,

    @Json(name = "gravatar_id")
    val gravatarID: String,

    val url: String,

    @Json(name = "avatar_url")
    val avatarURL: String
)

data class Payload (
    @Json(name = "push_id")
    val pushID: Long? = null,

    val size: Long? = null,

    @Json(name = "distinct_size")
    val distinctSize: Long? = null,

    val ref: String? = null,
    val head: String? = null,
    val before: String? = null,
    val commits: List<Commit>? = null,
    val action: String? = null,
    val number: Long? = null,

    @Json(name = "pull_request")
    val pullRequest: PullRequest? = null,

    @Json(name = "ref_type")
    val refType: String? = null,

    @Json(name = "master_branch")
    val masterBranch: String? = null,

    val description: String? = null,

    @Json(name = "pusher_type")
    val pusherType: String? = null,

    val comment: Comment? = null
)

data class Comment (
    val url: String,

    @Json(name = "pull_request_review_id")
    val pullRequestReviewID: Long,

    val id: Long,

    @Json(name = "node_id")
    val nodeID: String,

    @Json(name = "diff_hunk")
    val diffHunk: String,

    val path: String,
    val position: Long,

    @Json(name = "original_position")
    val originalPosition: Long,

    @Json(name = "commit_id")
    val commitID: String,

    @Json(name = "original_commit_id")
    val originalCommitID: String,

    val user: Owner,
    val body: String,

    @Json(name = "created_at")
    val createdAt: String,

    @Json(name = "updated_at")
    val updatedAt: String,

    @Json(name = "html_url")
    val htmlURL: String,

    @Json(name = "pull_request_url")
    val pullRequestURL: String,

    @Json(name = "author_association")
    val authorAssociation: String,

    @Json(name = "_links")
    val links: CommentLinks
)

data class CommentLinks (
    val self: HTML,
    val html: HTML,

    @Json(name = "pull_request")
    val pullRequest: HTML
)

data class HTML (
    val href: String
)

data class Owner (
    val login: String,
    val id: Long,

    @Json(name = "node_id")
    val nodeID: String,

    @Json(name = "avatar_url")
    val avatarURL: String,

    @Json(name = "gravatar_id")
    val gravatarID: String,

    val url: String,

    @Json(name = "html_url")
    val htmlURL: String,

    @Json(name = "followers_url")
    val followersURL: String,

    @Json(name = "following_url")
    val followingURL: String,

    @Json(name = "gists_url")
    val gistsURL: String,

    @Json(name = "starred_url")
    val starredURL: String,

    @Json(name = "subscriptions_url")
    val subscriptionsURL: String,

    @Json(name = "organizations_url")
    val organizationsURL: String,

    @Json(name = "repos_url")
    val reposURL: String,

    @Json(name = "events_url")
    val eventsURL: String,

    @Json(name = "received_events_url")
    val receivedEventsURL: String,

    val type: Type,

    @Json(name = "site_admin")
    val siteAdmin: Boolean
)

enum class Type(val value: String) {
    Organization("Organization"),
    User("User");

    companion object {
        public fun fromValue(value: String): Type = when (value) {
            "Organization" -> Organization
            "User"         -> User
            else           -> throw IllegalArgumentException()
        }
    }
}

data class Commit (
    val sha: String,
    val author: Author,
    val message: String,
    val distinct: Boolean,
    val url: String
)

data class Author (
    val email: String,
    val name: String
)

data class PullRequest (
    val url: String,
    val id: Long,

    @Json(name = "node_id")
    val nodeID: String,

    @Json(name = "html_url")
    val htmlURL: String,

    @Json(name = "diff_url")
    val diffURL: String,

    @Json(name = "patch_url")
    val patchURL: String,

    @Json(name = "issue_url")
    val issueURL: String,

    val number: Long,
    val state: String,
    val locked: Boolean,
    val title: String,
    val user: Owner,
    val body: String? = null,

    @Json(name = "created_at")
    val createdAt: String,

    @Json(name = "updated_at")
    val updatedAt: String,

    @Json(name = "closed_at")
    val closedAt: Any? = null,

    @Json(name = "merged_at")
    val mergedAt: Any? = null,

    @Json(name = "merge_commit_sha")
    val mergeCommitSHA: String? = null,

    val assignee: Any? = null,
    val assignees: List<Any?>,

    @Json(name = "requested_reviewers")
    val requestedReviewers: List<Any?>,

    @Json(name = "requested_teams")
    val requestedTeams: List<Any?>,

    val labels: List<Any?>,
    val milestone: Any? = null,

    @Json(name = "commits_url")
    val commitsURL: String,

    @Json(name = "review_comments_url")
    val reviewCommentsURL: String,

    @Json(name = "review_comment_url")
    val reviewCommentURL: String,

    @Json(name = "comments_url")
    val commentsURL: String,

    @Json(name = "statuses_url")
    val statusesURL: String,

    val head: Base,
    val base: Base,

    @Json(name = "_links")
    val links: PullRequestLinks,

    @Json(name = "author_association")
    val authorAssociation: String,

    val merged: Boolean? = null,
    val mergeable: Any? = null,
    val rebaseable: Any? = null,

    @Json(name = "mergeable_state")
    val mergeableState: String? = null,

    @Json(name = "merged_by")
    val mergedBy: Any? = null,

    val comments: Long? = null,

    @Json(name = "review_comments")
    val reviewComments: Long? = null,

    @Json(name = "maintainer_can_modify")
    val maintainerCanModify: Boolean? = null,

    val commits: Long? = null,
    val additions: Long? = null,
    val deletions: Long? = null,

    @Json(name = "changed_files")
    val changedFiles: Long? = null
)

data class Base (
    val label: String,
    val ref: String,
    val sha: String,
    val user: Owner,
    val repo: BaseRepo
)

data class BaseRepo (
    val id: Long,

    @Json(name = "node_id")
    val nodeID: String,

    val name: String,

    @Json(name = "full_name")
    val fullName: String,

    val owner: Owner,
    val private: Boolean,

    @Json(name = "html_url")
    val htmlURL: String,

    val description: String? = null,
    val fork: Boolean,
    val url: String,

    @Json(name = "forks_url")
    val forksURL: String,

    @Json(name = "keys_url")
    val keysURL: String,

    @Json(name = "collaborators_url")
    val collaboratorsURL: String,

    @Json(name = "teams_url")
    val teamsURL: String,

    @Json(name = "hooks_url")
    val hooksURL: String,

    @Json(name = "issue_events_url")
    val issueEventsURL: String,

    @Json(name = "events_url")
    val eventsURL: String,

    @Json(name = "assignees_url")
    val assigneesURL: String,

    @Json(name = "branches_url")
    val branchesURL: String,

    @Json(name = "tags_url")
    val tagsURL: String,

    @Json(name = "blobs_url")
    val blobsURL: String,

    @Json(name = "git_tags_url")
    val gitTagsURL: String,

    @Json(name = "git_refs_url")
    val gitRefsURL: String,

    @Json(name = "trees_url")
    val treesURL: String,

    @Json(name = "statuses_url")
    val statusesURL: String,

    @Json(name = "languages_url")
    val languagesURL: String,

    @Json(name = "stargazers_url")
    val stargazersURL: String,

    @Json(name = "contributors_url")
    val contributorsURL: String,

    @Json(name = "subscribers_url")
    val subscribersURL: String,

    @Json(name = "subscription_url")
    val subscriptionURL: String,

    @Json(name = "commits_url")
    val commitsURL: String,

    @Json(name = "git_commits_url")
    val gitCommitsURL: String,

    @Json(name = "comments_url")
    val commentsURL: String,

    @Json(name = "issue_comment_url")
    val issueCommentURL: String,

    @Json(name = "contents_url")
    val contentsURL: String,

    @Json(name = "compare_url")
    val compareURL: String,

    @Json(name = "merges_url")
    val mergesURL: String,

    @Json(name = "archive_url")
    val archiveURL: String,

    @Json(name = "downloads_url")
    val downloadsURL: String,

    @Json(name = "issues_url")
    val issuesURL: String,

    @Json(name = "pulls_url")
    val pullsURL: String,

    @Json(name = "milestones_url")
    val milestonesURL: String,

    @Json(name = "notifications_url")
    val notificationsURL: String,

    @Json(name = "labels_url")
    val labelsURL: String,

    @Json(name = "releases_url")
    val releasesURL: String,

    @Json(name = "deployments_url")
    val deploymentsURL: String,

    @Json(name = "created_at")
    val createdAt: String,

    @Json(name = "updated_at")
    val updatedAt: String,

    @Json(name = "pushed_at")
    val pushedAt: String,

    @Json(name = "git_url")
    val gitURL: String,

    @Json(name = "ssh_url")
    val sshURL: String,

    @Json(name = "clone_url")
    val cloneURL: String,

    @Json(name = "svn_url")
    val svnURL: String,

    val homepage: String? = null,
    val size: Long,

    @Json(name = "stargazers_count")
    val stargazersCount: Long,

    @Json(name = "watchers_count")
    val watchersCount: Long,

    val language: String,

    @Json(name = "has_issues")
    val hasIssues: Boolean,

    @Json(name = "has_projects")
    val hasProjects: Boolean,

    @Json(name = "has_downloads")
    val hasDownloads: Boolean,

    @Json(name = "has_wiki")
    val hasWiki: Boolean,

    @Json(name = "has_pages")
    val hasPages: Boolean,

    @Json(name = "forks_count")
    val forksCount: Long,

    @Json(name = "mirror_url")
    val mirrorURL: Any? = null,

    val archived: Boolean,

    @Json(name = "open_issues_count")
    val openIssuesCount: Long,

    val license: License? = null,
    val forks: Long,

    @Json(name = "open_issues")
    val openIssues: Long,

    val watchers: Long,

    @Json(name = "default_branch")
    val defaultBranch: String
)

data class License (
    val key: String,
    val name: String,

    @Json(name = "spdx_id")
    val spdxID: String? = null,

    val url: String? = null,

    @Json(name = "node_id")
    val nodeID: String
)

data class PullRequestLinks (
    val self: HTML,
    val html: HTML,
    val issue: HTML,
    val comments: HTML,

    @Json(name = "review_comments")
    val reviewComments: HTML,

    @Json(name = "review_comment")
    val reviewComment: HTML,

    val commits: HTML,
    val statuses: HTML
)

data class EventRepo (
    val id: Long,
    val name: String,
    val url: String
)

data class Gist (
    val url: String,

    @Json(name = "forks_url")
    val forksURL: String,

    @Json(name = "commits_url")
    val commitsURL: String,

    val id: String,

    @Json(name = "node_id")
    val nodeID: String,

    @Json(name = "git_pull_url")
    val gitPullURL: String,

    @Json(name = "git_push_url")
    val gitPushURL: String,

    @Json(name = "html_url")
    val htmlURL: String,

    val files: Map<String, File>,
    val public: Boolean,

    @Json(name = "created_at")
    val createdAt: String,

    @Json(name = "updated_at")
    val updatedAt: String,

    val description: String? = null,
    val comments: Long,
    val user: Any? = null,

    @Json(name = "comments_url")
    val commentsURL: String,

    val owner: Owner,
    val truncated: Boolean
)

data class File (
    val filename: String,
    val type: String,
    val language: String? = null,

    @Json(name = "raw_url")
    val rawURL: String,

    val size: Long
)

data class Meta (
    @Json(name = "verifiable_password_authentication")
    val verifiablePasswordAuthentication: Boolean,

    @Json(name = "github_services_sha")
    val githubServicesSHA: String,

    val hooks: List<String>,
    val git: List<String>,
    val pages: List<String>,
    val importer: List<String>
) {
    public fun toJson() = klaxon.toJsonString(this)

    companion object {
        public fun fromJson(json: String) = klaxon.parse<Meta>(json)
    }
}
